### PR TITLE
Remove duplicates from models #357

### LINKED ISSUE #357: Remove duplicates from models
move repeated code to base class

### PR DESCRIPTION
**Pull request by amper43**
fixes #357

### CODE REVIEW

File: analytics/.vscode/settings.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "python.pythonPath": "python",
+  "python.pythonPath": "python3",
```
Comment (jonyrock) [inline analytics/.vscode/settings.json]: not sure we should change it here, but I don`t care much

File: analytics/analytics/models/general_model.py [hunk 2]
```diff
@@ -47,16 +43,12 @@ def do_fit(self, dataframe: pd.DataFrame, segments: list) -> None:
...
-                deleted_pat = utils.get_interval(data, del_mid_index, self.state['WINDOW_SIZE'])
-                deleted_pat = utils.subtract_min_without_nan(segment_data)
-                del_conv_pat = scipy.signal.fftconvolve(deleted_pat, self.model_gen)
-                del_conv_list.append(max(del_conv_pat))
+        for segment in deleted_segments:
+            del_mid_index = segment.start + math.ceil(segment.length / 2)
+            deleted_pat = utils.get_interval(data, del_mid_index, self.state['WINDOW_SIZE'])
+            deleted_pat = utils.subtract_min_without_nan(deleted_pat)
+            del_conv_pat = scipy.signal.fftconvolve(deleted_pat, self.model_gen)
+            del_conv_list.append(max(del_conv_pat))
```
Comment (rozetko) [inline analytics/analytics/models/general_model.py]: I guess that `del_conv_pat` can be empty => `Error: max() arg is an empty sequence`

File: analytics/analytics/models/jump_model.py [hunk 3]
```diff
@@ -28,78 +28,39 @@ def __init__(self):
...
+        for segment in deleted_segments:
+            segment_cent_index = utils.find_parameters(segment.data, segment.start, 'jump')[0]
+            deleted_jump = utils.get_interval(data, segment_cent_index, self.state['WINDOW_SIZE'])
+            deleted_jump = utils.subtract_min_without_nan(deleted_jump)
+            del_conv_jump = scipy.signal.fftconvolve(deleted_jump, self.model_jump)
+            del_conv_list.append(max(del_conv_jump))
+
+        self._update_fiting_result(self.state, confidences, convolve_list, del_conv_list)
+        self.state['JUMP_HEIGHT'] = float(min(jump_height_list)) if jump_height_list else 1
+        self.state['JUMP_LENGTH'] = int(max(jump_length_list)) if jump_length_list else 1
```
Comment (rozetko) [inline analytics/analytics/models/jump_model.py]: Can `jump_height_list`/`jump_length_list` be empty?

File: analytics/analytics/models/model.py [hunk 4]
```diff
@@ -1,12 +1,33 @@
...
 
+class Segment(AttrDict):
+
+    __percent_of_nans = 0
+
+    def __init__(self, dataframe: pd.DataFrame, segment_map: dict):
+        self.update(segment_map)
+        self.start = utils.timestamp_to_index(dataframe, pd.to_datetime(self['from'], unit='ms'))
+        self.end = utils.timestamp_to_index(dataframe, pd.to_datetime(self['to'], unit='ms'))
+        self.data = dataframe['value'][self.start: self.end + 1]
```
Comment (rozetko) [inline analytics/analytics/models/model.py]: It would be better to check for dataframe length first

File: analytics/analytics/utils/common.py [hunk 5]
```diff
@@ -278,3 +278,8 @@ def cut_dataframe(data: pd.DataFrame) -> pd.DataFrame:
     if not np.isnan(data_min) and data_min > 0:
         data['value'] = data['value'] - data_min
     return data
+
+def get_min_max(array, default):
+    minimum = float(min(array)) if array else default
```
Comment (rozetko) [inline analytics/analytics/utils/common.py]: what if array is empty?

File: analytics/analytics/utils/common.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def segments_box(segments):
 
 def find_pattern(data: pd.Series, height: float, lenght: int, pattern_type: str) -> list:
     pattern_list = []
-    right_bound = len(data) - length - 1
+    right_bound = len(data) - lenght - 1
```
Comment (rozetko) [inline analytics/analytics/utils/common.py]: Would be better to rename parameter from `lenght` to `length` :)

**Pull request merged**